### PR TITLE
fix: server action & ssr on windows

### PR DESCRIPTION
**Pull request by himself65**
Fixes: https://github.com/dai-shi/waku/issues/192 and add SSR test cases

### CODE REVIEW

File: packages/waku/src/client.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ export const fetchRSC = cache(
     const options = {
       async callServer(actionId: string, args: unknown[]) {
         const response = fetch(
-          basePath + encodeInput(encodeURIComponent(actionId)),
+          basePath + encodeURI(encodeInput(encodeURIComponent(actionId))),
```
Comment (dai-shi) [inline packages/waku/src/client.ts:53]: This seems too ad-hoc. `input` is used for both components and actions.
I was hoping `encodeURIComponent` works, but if it doesn't escape enough, maybe we should use base64.
We can of course re-consider `encodeInput` implementation, but encoding too much is bad in DX.

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 2]
```diff
@@ -294,7 +297,10 @@ async function renderRSC(rr: RenderRequest): Promise<PipeableStream> {
     {},
     {
       get(_target, encodedId: string) {
-        const [filePath, name] = encodedId.split('#') as [string, string];
+        const [filePath, name] = normalizePath(encodedId).split('#') as [
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:300]: Shouldn't we normalize only the `filePath` part?

**Comment by himself65** [comment]
there is only windows prod mode now

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 3]
```diff
@@ -252,7 +244,7 @@ const transformRsfId = (prefixToRemove: string) =>
       for (let i = 0; i < lines.length; ++i) {
         const match = lines[i].match(
           new RegExp(
-            `^([0-9]+):{"id":"(?:file://)?${prefixToRemove}(.*?)"(.*)$`,
+            `^([0-9]+):{"id":"(?:file:\/\/\/)?${prefixToRemove}(.*?)"(.*)$`,
```
Comment (himself65) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:247]: where the bug comes from on Windows

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 4]
```diff
@@ -249,8 +244,8 @@ const transformRsfId = (prefixToRemove: string) =>
       for (let i = 0; i < lines.length; ++i) {
         const match = lines[i].match(
           new RegExp(
-            `^([0-9]+):{"id":"(?:file://)?${prefixToRemove}(.*?)"(.*)$`,
-          ),
+            `^([0-9]+):{"id":"(?:file://?/)?${prefixToRemove}(.*?)"(.*)$`
```
Comment (himself65) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:247]: On windows, there is one more extra `/`

**Comment by himself65** [comment]
@dai-shi I added the SSR test case, seems like no js environment doesn't work

File: e2e/ssr-basic.spec.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    test('no js environment should have first screen', async ({ browser }) => {
+      test.fixme();
+      const noJSContext = await browser.newContext({
+        javaScriptEnabled: false,
+      });
+      const noJSPage = await noJSContext.newPage();
+      await noJSPage.goto(`http://localhost:${port}/`);
+      await expect(noJSPage.getByTestId('app-name')).toHaveText('Waku');
+      await expect(noJSPage.getByTestId('count')).toHaveText('0');
```
Comment (himself65) [inline e2e/ssr-basic.spec.ts:75]: todo: fix this

File: packages/waku/src/client.ts [hunk 6]
```diff
@@ -49,13 +49,11 @@ export const fetchRSC = cache(
...
     const options = {
       async callServer(actionId: string, args: unknown[]) {
-        const response = fetch(
-          basePath + encodeInput(encodeURIComponent(actionId)),
-          {
-            method: 'POST',
-            body: await encodeReply(args),
-          },
-        );
+        const input = basePath + encodeInput(encodeURIComponent(actionId));
```
Comment (dai-shi) [inline packages/waku/src/client.ts:52]: In my terminology, `input` is a string passed to a function in `entries.ts`.
So, let's rename it:
```suggestion
        const reqPath = basePath + encodeInput(encodeURIComponent(actionId));
```

File: packages/waku/src/lib/builder.ts [hunk 7]
```diff
@@ -176,15 +176,19 @@ const buildServerBundle = async (
   const code = `export const resolveClientPath = (filePath, invert) => (invert ? ${JSON.stringify(
     Object.fromEntries(
       Object.entries(clientEntryFiles).map(([key, val]) => [
-        path.posix.join(config.rootDir, config.distDir, 'assets', key + '.js'),
+        normalizePath(
+          path.join(config.rootDir, config.distDir, 'assets', key + '.js'),
+        ),
```
Comment (dai-shi) [inline packages/waku/src/lib/builder.ts:181]: Nice.

File: packages/waku/src/lib/builder.ts [hunk 8]
```diff
@@ -279,8 +283,7 @@ const emitRscFiles = async (
           config.distDir,
           config.publicDir,
           config.rscPath,
-          // HACK to support windows filesystem
-          encodeInput(input).replaceAll('/', path.sep),
+          encodeInput(normalizePath(input)),
```
Comment (dai-shi) [inline packages/waku/src/lib/builder.ts:286]: Very nice.

File: packages/waku/src/lib/config.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@ import fs from 'node:fs';
 import path from 'node:path';
 
 import type { Config } from '../config.js';
+import { normalizePath } from 'vite';
```
Comment (dai-shi) [inline packages/waku/src/lib/config.ts:5]: This file is used in production. So, we can't depend on `vite`.

Maybe, we need to take `command` argument in `resolveConfig` function below, and only if it's `'dev'` , do `import('vite')`. We do this in various places.

File: packages/waku/src/lib/middleware/rsc.ts [hunk 10]
```diff
@@ -121,7 +121,10 @@ export function rsc<Context>(options: {
   return async (req, res, next) => {
     const config = await configPromise;
     const basePrefix = config.basePath + config.rscPath + '/';
-    const pathStr = req.url || '';
+    let pathStr = req.url || '';
+    if (pathStr.startsWith('/@fs/')) {
+      pathStr = pathStr.slice('/@fs/'.length);
+    }
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc.ts:127]: This is DEV-only, so only do it if `command==="dev"`.
Though, I'm not 100% sure if it's the right way to trim `/@fs` here.

File: packages/waku/src/lib/middleware/rsc/ssr.ts [hunk 11]
```diff
@@ -18,6 +18,7 @@ import { defineEntries } from '../../../server.js';
 import { ServerRoot } from '../../../client.js';
 import { renderRSC } from './worker-api.js';
 import { hasStatusCode } from './utils.js';
+import { normalizePath } from 'vite';
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/ssr.ts:21]: Same here. We can't depend vite in production mode.

File: packages/waku/src/lib/middleware/rsc/ssr.ts [hunk 12]
```diff
@@ -303,16 +304,20 @@ export const renderHtml = async <Context>(
             get(_target, name: string) {
               const file = filePath.slice(config.basePath.length);
               if (command === 'dev') {
-                const filePath = file.startsWith('@fs/')
-                  ? file.slice(3)
-                  : path.join(config.rootDir, config.srcDir, file);
+                const filePath = normalizePath(
+                  file.startsWith('@fs/')
+                    ? file.slice('@fs/'.length)
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/ssr.ts:309]: Is it intentional? `3 !== '@fs/'.length`

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 13]
```diff
@@ -18,6 +18,7 @@ import {
   defineEntries,
   runWithAsyncLocalStorage as runWithAsyncLocalStorageOrig,
 } from '../../../server.js';
+import { normalizePath } from 'vite';
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:21]: Same here again. If we need such a functionality in production, we need to create one for ourselves.

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 14]
```diff
@@ -249,7 +244,7 @@ const transformRsfId = (prefixToRemove: string) =>
       for (let i = 0; i < lines.length; ++i) {
         const match = lines[i].match(
           new RegExp(
-            `^([0-9]+):{"id":"(?:file://)?${prefixToRemove}(.*?)"(.*)$`,
+            `^([0-9]+):{"id":"(?:file://?/)?${prefixToRemove}(.*?)"(.*)$`,
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:247]: Hm, what was the case that this regex didn't work?

Comment (dai-shi) [reply packages/waku/src/lib/middleware/rsc/worker-impl.ts:247]: And, can it be this?
```suggestion
            `^([0-9]+):{"id":"(?:file:///?)?${prefixToRemove}(.*?)"(.*)$`,
```

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 15]
```diff
@@ -296,7 +291,9 @@ async function renderRSC(rr: RenderRequest): Promise<PipeableStream> {
       get(_target, encodedId: string) {
         const [filePath, name] = encodedId.split('#') as [string, string];
         const id = resolveClientEntry(
-          filePath,
+          filePath.startsWith('file:///')
```
[the remark this answers is not in the document]
Comment (dai-shi) [reply packages/waku/src/lib/middleware/rsc/worker-impl.ts:294]: I wonder if we can create a utility function. Maybe a separate pr.

File: packages/waku/src/lib/middleware/rsc/worker-impl.ts [hunk 16]
```diff
@@ -328,8 +325,10 @@ async function renderRSC(rr: RenderRequest): Promise<PipeableStream> {
         args = await decodeReply(body);
       }
     }
-    const [fileId, name] = actionId.split('#');
-    const filePath = path.join(config.rootDir, fileId!);
+    const [fileId, name] = actionId.split('#') as [string, string];
+    const filePath = normalizePath(
+      path.join(config.rootDir, normalizePath(fileId)),
+    );
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/worker-impl.ts:331]: You need to normalize twice?

**Review by dai-shi — CHANGES_REQUESTED** [review]
some of comments should be addressed before merging.

File: packages/waku/src/lib/middleware/rsc/ssr.ts (hunk 12 above)
Comment (himself65) [reply packages/waku/src/lib/middleware/rsc/ssr.ts]: mistake I think, auto-complete by copilot 🤔

File: packages/waku/src/lib/middleware/rsc.ts [hunk 17]
```diff
@@ -109,6 +109,9 @@ export function rsc<
     } catch (e) {
       // does not exist
     }
+    if (pathStr.startsWith('/@fs')) {
+      return null;
+    }
```
Comment (himself65) [inline packages/waku/src/lib/middleware/rsc.ts:114]: I dont know why pathStr could be `'/@fs/C://xxx/Counter.tsx` on Windows

File: packages/waku/src/lib/middleware/rsc/ssr.ts [hunk 18]
```diff
@@ -317,16 +317,21 @@ export const renderHtml = async <Context>(
...
+                    : path.join(config.rootDir, config.srcDir, file),
+                );
+                console.log('filePath', filePath);
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/ssr.ts:325]:
```suggestion
```

File: packages/waku/src/lib/middleware/rsc/utils.ts [hunk 19]
```diff
@@ -1,5 +1,14 @@
 // This file should not include Node specific code.
 
+// copy from Node.js source code
+const win32Sep = '\\';
+const posixSep = '/';
+
+// Refs: https://github.com/rollup/plugins/blob/d49bbe8dc5ec41157de5787c72c858f73be107ff/packages/pluginutils/src/normalizePath.ts
+export function normalizePath(filePath: string) {
+  return filePath.split(win32Sep).join(posixSep);
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc/utils.ts:9]: That's simple.
I'd inline them, as it's fairly readable.
```suggestion
  return filePath.split('\\').join('/');
```

This doesn't work if posix path contains backslashes. Can't be helped.

**Review by dai-shi — COMMENTED** [review]
Looks good. Please see comments.

**Review by dai-shi — APPROVED** [review]

File: packages/waku/src/lib/middleware/rsc/ssr.ts [hunk 20]
```diff
@@ -319,7 +319,9 @@ export const renderHtml = async <Context>(
               if (command === 'dev') {
                 const filePath = normalizePath(
                   file.startsWith('@fs/')
-                    ? file.slice('@fs/'.length)
+                    // FIXME This is ugly. We need to refactor it.
+                    // remove '@fs'(3) on Unix and '@fs/'(4) on Windows
+                    ? file.slice(2 + path.sep.length)
```
Comment (himself65) [inline packages/waku/src/lib/middleware/rsc/ssr.ts:324]: I just found that `@fs/User/path/to/file` on Unix, but `@fs/C://path/to/file` on Windows

Is this kind of **feature** on Vite?

**Review by dai-shi — COMMENTED** [review]
Now, I need to resolve conflicts in my PRs.

**Comment by dai-shi** [comment]
Ah, this PR broke `npm run examples:dev:03_promise` and `05` too. And also `07`.

**Pull request merged**